Add device triggers for Hue remotes

azogue opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Following #33277, this PR adds device triggers for Hue dimmer switch and Hue Tap remotes.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml

```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [x] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

azogue committed:
Store device_registry entry id in HueEvent

so it can be retrieved with that key when using device triggers

azogue committed:
Add device_trigger for hue_event from hue remotes

* supporting Hue dimmer switch & Hue Tap
* State mapping and strings are copied from deCONZ

balloob reviewed (approved):
Niiiiice! Ok to merge when test added.

azogue committed:
refactor mock_bridge for hue tests

and also share `setup_bridge_for_sensors`
for test_sensor_base and test_device_trigger.

azogue committed:
Add tests for device triggers with hue remotes

azogue commented:
Some questions, @balloob 

1. I just copied the strings from deconz, and `translations_develop` just generates the `en.json` entries, but, should I manually copy the used keys & strings for the rest of the translations? It seems silly to send them to translate when we already have the translations for deconz, right?

2. About the triggers for the dimmer switch: in normal conditions only the 'button releases' events are going to appear, because of the polling, so maybe only those should be added, cutting in half the number of triggers, but avoiding user confusion. What do you think?
  I didn't remove them because, even with the polling, a button initial press could be got in the middle of an update, and in that case the initial press event would be fired (and the button release event in the next scan_interval)

balloob commented:
1. We don't currently have a way set up to share translations.

2. Let's keep it as limited as possible for first release maybe. If things can end up firing twice, that would be pretty bad.

azogue commented:
> 2\. Let's keep it as limited as possible for first release maybe. If things can end up firing twice, that would be pretty bad.

Oh no, no. Not twice, as each event/trigger is different. Thing is that for dimmers there is an event for button press and another for button released. As this is not pushing, unless you maintain the finger pressed while HA updates, the first event/trigger is never caught, and that's why I was suggesting to remove these. I'll do it.

balloob commented:
If you do it fast, it can go in the beta :)

azogue committed:
Remove some triggers

balloob commented:
Adding translations won't help, they are automatically written by Lokalise.

balloob commented on homeassistant/components/hue/strings.json:
```diff
@@ -27,5 +27,22 @@
+    },
+    "trigger_type": {
+      "remote_button_long_release": "\"{subtype}\" button released after long press",
+      "remote_button_short_press": "\"{subtype}\" button pressed",
```
I think that since we poll, we should drop this one too?

azogue replied:
That one is assigned to the Tap switch (without press/release), as copied from deconz. Not in use with the dimmer

azogue commented:
> Adding translations won't help, they are automatically written by Lokalise.

Oh, sorry. I did it for Spanish in a new integration I did recently. I'll revert it

MartinHjelmare commented on tests/components/hue/conftest.py:
```diff
@@ -1,11 +1,95 @@
+        allow_groups=False,
+        api=Mock(),
+        reset_jobs=[],
+        spec=hue.HueBridge,
```
We should not mock our own code. We should patch and mock the library code. That will make the tests more robust.

MartinHjelmare commented on tests/components/hue/conftest.py:
```diff
@@ -1,11 +1,95 @@
+    if hostname is None:
+        hostname = "mock-host"
+    hass.config.components.add(hue.DOMAIN)
+    config_entry = config_entries.ConfigEntry(
```
Use `MockConfigentry` from tests/common.py instead.

MartinHjelmare commented on tests/components/hue/conftest.py:
```diff
@@ -1,11 +1,95 @@
+        system_options={},
+    )
+    mock_bridge.config_entry = config_entry
+    hass.data[hue.DOMAIN] = {config_entry.entry_id: mock_bridge}
```
It's better to set up the integration than mocking integration code. Patch the library I/O as needed.

MartinHjelmare commented on tests/components/hue/test_device_trigger.py:
```diff
@@ -0,0 +1,169 @@
+    mock_bridge.mock_sensor_responses.append(new_sensor_response)
+
+    # Force updates to run again
+    await mock_bridge.sensor_manager.coordinator.async_refresh()
```
Patch time and fire a time changed event.

MartinHjelmare reviewed (commented):
Comments for the future.

johansmitsnl commented:
@azogue does this use the same refresh delay of 5s like the normal sensors use or are these faster?

azogue commented:
> @azogue does this use the same refresh delay of 5s like the normal sensors use or are these faster?

Hi @johansmitsnl, yes it does :(

The call to the Hue bridge is shared for all sensors and remotes (so no extra calls are made), but the `scan_interval` of 5s is not touched, so same refresh delay.

If you need instant response I suggest moving those devices to another ZigBee concentrator with _pushing_ support (like the ConBee usb stick). I did it myself and I'm very happy with the results. 

Here it can be useful to trigger automations in HA with _hold_ button presses, while maintaining the Hue app config to direct control of hue lights or groups with simple presses.

Also, this implementation is mirrored from the deCONZ, so if you someday make the jump between zigbee hubs, the only thing that would change is the event name (`hue_event` -> `deconz_event`)

johansmitsnl commented:
@azogue thanks for the quick reply.
